src/libopensc/pkcs15-sc-hsm.c

```diff
@@ -334,10 +418,332 @@ int sc_pkcs15emu_sc_hsm_decode_cvc(sc_pkcs15_card_t * p15card,
+
+	card = p15card->card;
+
+	format = calloc(1, sizeof(*format));
+	if (format == NULL) {
+		r = -1;
+		goto err;
+	}
+
+	sc_asn1_sc_hsm_pka_data_init(card->ctx, &format->data, pka);
+
+	sc_copy_asn1_entry(c_asn1_sc_hsm_pka_old_format, format->seq);
+	format->seq[0] = format->data.asn1_public_key_req[0];
+	format->seq[1] = format->data.asn1_device_cvc.asn1_cvc[0];
+	format->seq[2] = format->data.asn1_dica_cvc.asn1_cvc[0];
+
+	r = sc_asn1_decode(p15card->card->ctx, format->seq, *buf, *buflen,
+			buf, buflen);
+	LOG_TEST_GOTO_ERR(card->ctx, r,
+		"Could not decode ASN.1 for public key file's old format");
+
+	r = 0;
+	/* fall-through */
+
+err:
+	if (format != NULL) {
+		free(format);
+		format = NULL;
+	}
+	return r;
+}
+
+/*
+ * For SmartCard HSMs, this is the newer format for registering a public key
+ * for public key authentication.
+ *
+ * @param buf: *buf should point to the first tag after the sequence tag
+ *
+ * 1.3.6.1.4.1.24991 is the CardContact organization
+ * The 4.3.1 part is from inspecting their exported public key, but it doesn't
+ * seem to be publicly registered.
+ *
+ * SEQUENCE (0x30)
+ *   OID (0x6) 1.3.6.1.4.1.24991.4.3.1
+ *   Application 1 (0x61)
+ *      device CVC (0x7F21)
+ *   Application 2 (0x62)
+ *      device issuer CA CVC (0x7F21)
+ *   Application 3 (0x63)
+ *      authenticatedrequest for public key details (0x67)
+ */
+static int decode_pka_new_format(sc_pkcs15_card_t *p15card,
+							const u8 **buf, size_t *buflen,
+							sc_cvc_pka_t *pka)
+{
+	int r;
+	sc_card_t *card;
+	struct sc_asn1_sc_hsm_pka_new_format *format = NULL;
+
+	card = p15card->card;
+
+	format = calloc(1, sizeof(*format));
+	if (format == NULL) {
+		r = -1;
+		goto err;
+	}
+
+	sc_asn1_sc_hsm_pka_data_init(card->ctx, &format->data, pka);
+
+	sc_copy_asn1_entry(c_asn1_sc_hsm_pka_new_format, format->seq);
+	sc_format_asn1_entry(&format->seq[0], &format->oid, NULL, 0);
+	sc_format_asn1_entry(&format->seq[1],
+			format->data.asn1_dica_cvc.asn1_cvc, NULL, 0);
+	sc_format_asn1_entry(&format->seq[2],
+			format->data.asn1_device_cvc.asn1_cvc, NULL, 0);
+	sc_format_asn1_entry(&format->seq[3],
+			format->data.asn1_public_key_req, NULL, 0);
+
+	r = sc_asn1_decode(p15card->card->ctx, format->seq, *buf, *buflen,
+			buf, buflen);
+	LOG_TEST_GOTO_ERR(card->ctx, r,
+		"Could not decode ASN.1 for public key file's new format");
+
+	if (sc_compare_oid(&format->oid, &sc_hsm_public_key_oid) == 0) {
+		/* sc_dump_oid uses static memory */
+		sc_log(p15card->card->ctx, "OID %s did not match expected value",
+			   sc_dump_oid(&format->oid));
+		r = -1;
+		goto err;
+	}
+
+	r = 0;
+	/* fall-through */
+
+err:
+	if (format != NULL) {
+		free(format);
+		format = NULL;
+	}
+	return r;
+}
+
+/*
+ * @param pka: will be overwritten, should be uninitialized or memset to 0
+ */
+int sc_pkcs15emu_sc_hsm_decode_pka(sc_pkcs15_card_t *p15card,
+											const u8 **buf, size_t *buflen,
+											sc_cvc_pka_t *pka)
```

Comment:
your indent setting looks very weired

Reply:
sorry not used to \t tabs :laughing: i changed the functions to overflow with a single indent